### PR TITLE
unit: don't assume the NameOwner released if we fail to GetNameOwner

### PR DESCRIPTION
**Pull request by lclgo**

### CODE REVIEW

**Comment by Geass-LL** [comment]
When dbus fails to reply the "GetNameOwner" message, it doesn't mean this dbus connection has been closed, maybe dbus is too busy to reply it. But we used to assume this dbus connection has been closed, and will kill the dbus type service by SIGTERM later, this is dangerous.

You can easily reproduce this problem by `gdb attach dbus-daemon`, and run `systemctl daemon-reload`, then wait for about 1 min to find out NetworkManager, systemd-logind, polkit... have been killed by systemd.

**Comment by yuwata** [comment]
Please extend the commit message:
- prefix commit title with `core:` or `core/unit:`, rather than `unit:`,
- please include the explanation in https://github.com/systemd/systemd/pull/27763#issuecomment-1560421420,

Please add a testcase.

**Comment by Geass-LL** [comment]
The comment and title have been added, but I don't know how to add a testcase. How can I hang dbus-daemon in the testcase?

**Review by yuwata — APPROVED** [review]
LGTM.

> but I don't know how to add a testcase. How can I hang dbus-daemon in the testcase?

OK. Let's consider that later.

File: src/core/unit.c [hunk 1]
```diff
@@ -3456,6 +3456,7 @@ static int get_name_owner_handler(sd_bus_message *message, void *userdata, sd_bu
                         log_unit_error_errno(u, r,
                                              "Unexpected error response from GetNameOwner(): %s",
                                              bus_error_message(e, r));
+                        return 0;
```
Comment (poettering) [inline src/core/unit.c]: hmm, I am not sure if I am onboard with this. I think a bus service on a broker that doesn't work can be assumed to be dead. 

It appears to me as if this change makes things worse: we'll have a service we consider running even though we cannot talk to it.

Can you elaborate on the specific issue you saw here, and what kind of error you ran into specifically?

**Comment by poettering** [comment]
i mean, I'd be fine if we do the method call with an extra-long timeout, but I am very sure we should eventually assume the service to be dead if we cannot get a reply.

**Comment by Geass-LL** [comment]
> Can you elaborate on the specific issue you saw here, and what kind of error you ran into specifically?

I encountered this problem when running `systemctl daemon-reload` while there are many messages transporting from broker to NetworkManager/firewalld/nmcli... And I got this error:

```
May 20 23:01:25 sinon systemd[1]: systemd-logind.service: Unexpected error response from GetNameOwner(): Connection terminated
May 20 23:01:25 sinon systemd[1]: firewalld.service: Unexpected error response from GetNameOwner(): Connection terminated
May 20 23:01:25 sinon systemd[1]: polkit.service: Unexpected error response from GetNameOwner(): Connection terminated
May 20 23:01:25 sinon sysmonitor[21474]: sysmonitor[21474]: comm:systemd exe:systemd[1](parent comm:swapper/0 parent exe:[0]) send SIGTERM to comm:systemd-logind exe:systemd-logind[2610100].
May 20 23:01:25 sinon sysmonitor[21474]: sysmonitor[21474]: comm:systemd exe:systemd[1](parent comm:swapper/0 parent exe:[0]) send SIGTERM to comm:firewalld exe:python3.9[2526119].
May 20 23:01:25 sinon systemd-logind[2610100]: Failed to start session scope session-12327.scope: Message recipient disconnected from message bus without replying
May 20 23:01:25 sinon sysmonitor[21474]: sysmonitor[21474]: comm:systemd exe:systemd[1](parent comm:swapper/0 parent exe:[0]) send SIGTERM to comm:polkitd exe:polkitd[2509416].
May 20 23:01:25 sinon sshd[328652]: pam_systemd(sshd:session): Failed to create session: Message recipient disconnected from message bus without replying
```

sysmonitor is a daemon we used to monitor SIGTERM. After this `Unexpected error response from GetNameOwner`, the  coresponding dbus type service is killed by systemd. This is safe for systemd-logind, because it has configured `Restart=always`, and it can be restarted later. But if NetworkManager is killed, it will not be restarted. In the view of NetworkManager, it dosn't do anything wrong, but killed.

> It appears to me as if this change makes things worse: we'll have a service we consider running even though we cannot talk to it.

Although we cann't talk to it, we are monitoring its main process. We should leave the choice to service itself. i.e. decides to exit if it loses connection with dbus. But shouldn't just kill them.

**Comment by poettering** [comment]
sorry, but I think taping over issues is the wrong approach. I am fine with prolonging the time-out, but just getting rid of the check is wrong: if a service is not reachable via dbus it's a problem, and we need to handle it, and it doesn't matter that in some sense of the word the service is still running. A service that provides IPC and doesn't respond to IPC *should* be killed.

Sorry, but this is not the right approach, I am very very sure.

**Comment by Geass-LL** [comment]
I have token the advice from @poettering and update the PR to change the timeout to 90s. But there is a strange problem, although I have set the timeout to 90s, the get_name_owner_handler still triggered after 25s. While if I set the timeout to 5s, the get_name_owner_handler triggers fine. I don't know if this is a feature or bug, I'm still debugging.

**Comment by Geass-LL** [comment]
> I have token the advice from @poettering and update the PR to change the timeout to 90s. But there is a strange problem, although I have set the timeout to 90s, the get_name_owner_handler still triggered after 25s. While if I set the timeout to 5s, the get_name_owner_handler triggers fine. I don't know if this is a feature or bug, I'm still debugging.

It's because in `unit_install_bus_match()`, systemd creates a `AddMatch` async call first, the timeout is 25s. When this `AddMatch` call timedout, systemd will trigger `add_match_callback`, and in this function it use `bus_enter_closing()` close the connection. Before closing the connection, systemd will explicitly call `process_closing_reply_callback()`, and this will call `get_name_owner_handler` eventually.

**Pull request closed without merging**